File: java/core/src/core/util/Base16.java
/**
 * Author: Timothy Prepscius
 * License: GPLv3 Affero + keep my name in the code!
 */

package core.util;

public class Base16 
{
	static String decoder = "0123456789abcdef";
	static byte[] encoder = null;
	
	static void initialize ()
	{
		if (encoder != null)
			return;
		
		encoder = new byte[0xFF];
		for (int i=0; i<decoder.length(); ++i)
			encoder[decoder.charAt(i)] = (byte)i;
	}
	
	public static byte[] decode (String value)
	{
		initialize();
		
		byte[] bytes = new byte[value.length() / 2];
		
		int i=0, j=0;
		while (i<value.length())
		{
			int nibbleHi = (int)encoder[(int)value.charAt(i++)];
			int nibbleLo = (int)encoder[(int)value.charAt(i++)];
			
			bytes[j++] = (byte) ((nibbleHi << 4) | (nibbleLo));
		}
		
		return bytes;
	}

	public static String encode (byte[] bytes)
	{
		String result = "";
		
		int i=0;
		while (i<bytes.length)
		{
			int b = bytes[i++] & 0xFF;
			int nibbleLo = b & 0x0F;
			int nibbleHi = b >> 4;
		
			result += decoder.charAt(nibbleHi);
			result += decoder.charAt(nibbleLo);
		}
		
		return result;
	}
}

File: cpp/core/mailiverse/core/crypt/Base16.cpp
/**
 * Author: Timothy Prepscius
 * License: BSD + keep my name in the code!
 */

#include "Base16.h"
#include <botan/botan.h>
#include "CryptorUtil.h"

namespace mailiverse {
namespace core {
namespace crypt {

std::string Base16::encode (const Block &block)
{
	Botan::Pipe pipe(new Botan::Hex_Encoder(Botan::Hex_Encoder::Lowercase));
	pipe.process_msg(block.data(), block.size());
	Botan::SecureVector<Botan::byte> sv;

	return pipe.read_all_as_string();
}

Block Base16::decode (const std::string &block)
{
	Botan::Pipe pipe(new Botan::Hex_Encoder(Botan::Hex_Encoder::Lowercase));
	pipe.process_msg((Botan::byte *)block.data(), block.size());

	return read_all_as_block(&pipe);
}

Block Base16::encodeBytes (const Block &block)
{
	Botan::Pipe pipe(new Botan::Hex_Encoder(Botan::Hex_Encoder::Lowercase));
	pipe.process_msg(block.data(), block.size());
	Botan::SecureVector<Botan::byte> sv;

	return read_all_as_block(&pipe);
}

Block Base16::decodeBytes (const Block &block)
{
	Botan::Pipe pipe(new Botan::Hex_Encoder(Botan::Hex_Encoder::Lowercase));
	pipe.process_msg((Botan::byte *)block.data(), block.size());

	return read_all_as_block(&pipe);
}

} /* namespace crypt */
} /* namespace core */
} /* namespace mailiverse */
